File: rust/src/bin/latency_calculator.rs
use anyhow::Result;
use clap::Parser;
use futures::StreamExt;
use futures_util::{SinkExt, StreamExt as FuturesStreamExt};
use helius_laserstream::{
    grpc::{SubscribeRequest, SubscribeRequestFilterBlocks},
    subscribe, LaserstreamConfig,
};
use reqwest::Client;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use tokio::time;
use tokio_tungstenite::{connect_async, tungstenite::protocol::Message};
use helius_laserstream::grpc::SubscribeRequestFilterBlocksMeta;

#[derive(Parser)]
#[command(name = "latency-calculator")]
#[command(about = "Calculate average latency for RPC or gRPC over specified number of blocks")]
struct Args {
    #[arg(long, value_enum, help = "Method to test: rpc, grpc, or websocket")]
    method: Method,

    #[arg(long, help = "Endpoint URL")]
    endpoint: String,

    #[arg(long, help = "API key (for gRPC)")]
    api_key: Option<String>,

    #[arg(long, help = "Number of blocks to calculate average latency")]
    blocks: u64,

    #[arg(long, help = "Verbose logging")]
    verbose: bool,
}

#[derive(clap::ValueEnum, Clone, Debug)]
enum Method {
    Rpc,
    Grpc,
    Websocket,
}

#[derive(Debug, Clone)]
struct LatencyMeasurement {
    slot: u64,
    block_time: i64,
    received_time: i64,
    latency_ms: i64,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();

    println!("🚀 Latency Calculator");
    println!("Method: {:?}", args.method);
    println!("Endpoint: {}", args.endpoint);
    println!("Target blocks: {}", args.blocks);
    println!();

    let measurements = match args.method {
        Method::Rpc => measure_rpc_latency(&args).await?,
        Method::Grpc => measure_grpc_latency(&args).await?,
        Method::Websocket => measure_websocket_latency(&args).await?,
    };

    print_results(&measurements, &args);

    Ok(())
}

async fn measure_rpc_latency(args: &Args) -> Result<Vec<LatencyMeasurement>> {
    let client = Client::builder().timeout(Duration::from_secs(10)).build()?;
    let mut measurements = Vec::new();
    let mut processed_blocks = 0u64;

    println!("📡 Starting RPC latency measurement...");
    println!("Slot       | Block Time    | Received Time | Latency   | Status");
    println!("{}", "-".repeat(70));

    let mut last_slot = get_latest_slot(&client, &args.endpoint).await?;

    while processed_blocks < args.blocks {
        match get_latest_slot(&client, &args.endpoint).await {
            Ok(current_slot) => {
                if current_slot > last_slot {
                    // Process the new slot
                    match get_block_time(&client, &args.endpoint, current_slot).await {
                        Ok(Some(block_time)) => {
                            let received_time = SystemTime::now()
                                .duration_since(UNIX_EPOCH)?
                                .as_millis() as i64;

                            let latency_ms = received_time - (block_time * 1000);

                            // Filter out unrealistic latencies
                            if latency_ms > 0 && latency_ms < 10000 {
                                let measurement = LatencyMeasurement {
                                    slot: current_slot,
                                    block_time,
                                    received_time,
                                    latency_ms,
                                };

                                let status = get_latency_status(latency_ms);
                                
                                println!(
                                    "{:<10} | {:<12} | {:<12} | {:<9}ms | {}",
                                    current_slot,
                                    block_time,
                                    received_time / 1000,
                                    latency_ms,
                                    status
                                );

                                measurements.push(measurement);
                                processed_blocks += 1;

                                if args.verbose {
                                    println!("Progress: {}/{} blocks processed", processed_blocks, args.blocks);
                                }
                            }
                        }
                        Ok(None) => {
                            if args.verbose {
                                println!("Block time not available for slot {}", current_slot);
                            }
                        }
                        Err(e) => {
                            if args.verbose {
                                println!("Error getting block time for slot {}: {}", current_slot, e);
                            }
                        }
                    }
                    last_slot = current_slot;
                }
            }
            Err(e) => {
                if args.verbose {
                    eprintln!("Error getting latest slot: {}", e);
                }
            }
        }

        time::sleep(Duration::from_millis(500)).await;
    }

    Ok(measurements)
}

async fn measure_grpc_latency(args: &Args) -> Result<Vec<LatencyMeasurement>> {
    let api_key = args
        .api_key
        .clone()
        .or_else(|| std::env::var("HELIUS_API_KEY").ok())
        .ok_or_else(|| anyhow::anyhow!("API key required for gRPC method"))?;

    let config = LaserstreamConfig {
        api_key,
        endpoint: args.endpoint.parse()?,
        ..Default::default()
    };

    let mut request = SubscribeRequest::default();
    
    request.blocks_meta.insert(
        "all".to_string(),
        SubscribeRequestFilterBlocksMeta::default(),
    );

    let (stream, _handle) = subscribe(config, request);
    futures::pin_mut!(stream);

    let mut measurements = Vec::new();
    let mut processed_blocks = 0u64;

    println!("📡 Starting gRPC latency measurement...");
    println!("Slot       | Block Time    | Received Time | Latency   | Status");
    println!("{}", "-".repeat(70));

    while processed_blocks < args.blocks {
        if let Some(result) = stream.next().await {
            match result {
                Ok(update) => {
                    let received_time = SystemTime::now()
                        .duration_since(UNIX_EPOCH)?
                        .as_millis() as i64;

                    if let Some(helius_laserstream::grpc::subscribe_update::UpdateOneof::BlockMeta(block)) = update.update_oneof {
                        let slot = block.slot;
                        
                        if let Some(bt) = block.block_time {
                            let block_time = bt.timestamp;
                            let latency_ms = received_time - (block_time * 1000);

                            // Filter out unrealistic latencies
                            if latency_ms > 0 && latency_ms < 10000 {
                                let measurement = LatencyMeasurement {
                                    slot,
                                    block_time,
                                    received_time,
                                    latency_ms,
                                };

                                let status = get_latency_status(latency_ms);
                                
                                println!(
                                    "{:<10} | {:<12} | {:<12} | {:<9}ms | {}",
                                    slot,
                                    block_time,
                                    received_time / 1000,
                                    latency_ms,
                                    status
                                );

                                measurements.push(measurement);
                                processed_blocks += 1;

                                if args.verbose {
                                    println!("Progress: {}/{} blocks processed", processed_blocks, args.blocks);
                                }
                            }
                        }
                    }
                }
                Err(e) => {
                    if args.verbose {
                        eprintln!("gRPC stream error: {}", e);
                    }
                }
            }
        }
    }

    Ok(measurements)
}

async fn measure_websocket_latency(args: &Args) -> Result<Vec<LatencyMeasurement>> {
    let mut measurements = Vec::new();
    let mut processed_blocks = 0u64;

    println!("📡 Starting WebSocket latency measurement...");
    println!("Slot       | Block Time    | Received Time | Latency   | Status");
    println!("{}", "-".repeat(70));

    // Convert HTTP(S) URL to WebSocket URL
    let ws_url = if args.endpoint.starts_with("https://") {
        args.endpoint.replace("https://", "wss://")
    } else if args.endpoint.starts_with("http://") {
        args.endpoint.replace("http://", "ws://")
    } else if !args.endpoint.starts_with("ws://") && !args.endpoint.starts_with("wss://") {
        format!("wss://{}", args.endpoint)
    } else {
        args.endpoint.clone()
    };

    let (ws_stream, _) = connect_async(&ws_url).await?;
    let (mut write, mut read) = ws_stream.split();

    // Subscribe to block notifications
    let subscription = json!({
        "jsonrpc": "2.0",
        "id": 1,
        "method": "blockSubscribe",
        "params": [
            "all",
            {
                "commitment": "processed",
                "encoding": "json",
                "transactionDetails": "none",
                "rewards": false
            }
        ]
    });

    write.send(Message::Text(subscription.to_string())).await?;

    // Handle subscription confirmation and block notifications
    let mut subscription_confirmed = false;
    while processed_blocks < args.blocks {
        let timeout = tokio::time::sleep(Duration::from_secs(30));
        tokio::pin!(timeout);
        
        tokio::select! {
            msg_result = read.next() => {
                if let Some(msg) = msg_result {
                    match msg? {
                Message::Text(text) => {
                    let received_time = SystemTime::now()
                        .duration_since(UNIX_EPOCH)?
                        .as_millis() as i64;

                    if let Ok(json_msg) = serde_json::from_str::<Value>(&text) {
                        if args.verbose {
                            println!("Received WebSocket message: {}", serde_json::to_string_pretty(&json_msg).unwrap_or_else(|_| "Invalid JSON".to_string()));
                        }
                        // Check if this is a block notification
                        if let Some(params) = json_msg.get("params") {
                            if let Some(result) = params.get("result") {
                                if let Some(value) = result.get("value") {
                                    if let Some(block) = value.get("block") {
                                        let slot = value.get("slot")
                                            .and_then(|s| s.as_u64())
                                            .unwrap_or_else(|| {
                                                block.get("parentSlot")
                                                    .and_then(|s| s.as_u64())
                                                    .unwrap_or_default() + 1
                                            });
                                        
                                        if let Some(block_time) = block.get("blockTime").and_then(|bt| bt.as_i64()) {
                                            let latency_ms = received_time - (block_time * 1000);

                                            // Filter out unrealistic latencies
                                            if latency_ms > 0 && latency_ms < 10000 {
                                                let measurement = LatencyMeasurement {
                                                    slot,
                                                    block_time,
                                                    received_time,
                                                    latency_ms,
                                                };

                                                let status = get_latency_status(latency_ms);
                                                
                                                println!(
                                                    "{:<10} | {:<12} | {:<12} | {:<9}ms | {}",
                                                    slot,
                                                    block_time,
                                                    received_time / 1000,
                                                    latency_ms,
                                                    status
                                                );

                                                measurements.push(measurement);
                                                processed_blocks += 1;

                                                if args.verbose {
                                                    println!("Progress: {}/{} blocks processed", processed_blocks, args.blocks);
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        } else if json_msg.get("result").is_some() {
                            // Subscription confirmation
                            subscription_confirmed = true;
                            if args.verbose {
                                println!("WebSocket subscription confirmed");
                            }
                        }
                    }
                }
                Message::Close(_) => {
                    println!("WebSocket connection closed");
                    break;
                }
                _ => {}
                    }
                } else {
                    println!("WebSocket connection ended");
                    break;
                }
            }
            _ = &mut timeout => {
                if !subscription_confirmed {
                    return Err(anyhow::anyhow!("WebSocket subscription timeout"));
                } else {
                    println!("No new blocks received in 30 seconds, continuing...");
                }
            }
        }
    }

    Ok(measurements)
}

fn print_results(measurements: &[LatencyMeasurement], args: &Args) {
    if measurements.is_empty() {
        println!("❌ No measurements collected");
        return;
    }

    let latencies: Vec<i64> = measurements.iter().map(|m| m.latency_ms).collect();
    let total: i64 = latencies.iter().sum();
    let avg = total as f64 / latencies.len() as f64;
    
    let mut sorted_latencies = latencies.clone();
    sorted_latencies.sort();
    
    let min = *sorted_latencies.first().unwrap();
    let max = *sorted_latencies.last().unwrap();
    let median = sorted_latencies[sorted_latencies.len() / 2];
    let p95_idx = (sorted_latencies.len() as f64 * 0.95) as usize;
    let p95 = sorted_latencies[p95_idx.min(sorted_latencies.len() - 1)];
    let p99_idx = (sorted_latencies.len() as f64 * 0.99) as usize;
    let p99 = sorted_latencies[p99_idx.min(sorted_latencies.len() - 1)];

    // Count performance categories
    let excellent = latencies.iter().filter(|&&l| l < 500).count();
    let good = latencies.iter().filter(|&&l| l >= 500 && l < 1000).count();
    let fair = latencies.iter().filter(|&&l| l >= 1000 && l < 2000).count();
    let slow = latencies.iter().filter(|&&l| l >= 2000).count();

    println!();
    println!("📊 Latency Results Summary");
    println!("{}", "=".repeat(50));
    println!("Method:             {:?}", args.method);
    println!("Endpoint:           {}", args.endpoint);
    println!("Blocks processed:   {}", measurements.len());
    println!("Average latency:    {:.1}ms", avg);
    println!("Min latency:        {}ms", min);
    println!("Max latency:        {}ms", max);
    println!("Median latency:     {}ms", median);
    println!("95th percentile:    {}ms", p95);
    println!("99th percentile:    {}ms", p99);
    println!();

    println!("⚡ Performance Distribution:");
    println!("🟢 Excellent (<500ms):   {}/{} ({:.1}%)", 
        excellent, measurements.len(), (excellent as f64 / measurements.len() as f64) * 100.0);
    println!("🟡 Good (500-1000ms):    {}/{} ({:.1}%)", 
        good, measurements.len(), (good as f64 / measurements.len() as f64) * 100.0);
    println!("🟠 Fair (1000-2000ms):   {}/{} ({:.1}%)", 
        fair, measurements.len(), (fair as f64 / measurements.len() as f64) * 100.0);
    println!("🔴 Slow (>2000ms):       {}/{} ({:.1}%)", 
        slow, measurements.len(), (slow as f64 / measurements.len() as f64) * 100.0);
    println!();

    println!("🎯 Overall Assessment:");
    if avg < 500.0 {
        println!("✅ EXCELLENT - Very fast latency!");
    } else if avg < 1000.0 {
        println!("🟡 GOOD - Acceptable latency for most use cases");
    } else if avg < 2000.0 {
        println!("🟠 FAIR - Moderate latency, consider optimization");
    } else {
        println!("🔴 SLOW - High latency, investigate network/provider issues");
    }
}

fn get_latency_status(latency_ms: i64) -> &'static str {
    if latency_ms < 500 {
        "🟢 EXCELLENT"
    } else if latency_ms < 1000 {
        "🟡 GOOD"
    } else if latency_ms < 2000 {
        "🟠 FAIR"
    } else {
        "🔴 SLOW"
    }
}

async fn get_latest_slot(client: &Client, rpc_url: &str) -> Result<u64> {
    let request = json!({
        "jsonrpc": "2.0",
        "id": 1,
        "method": "getSlot",
        "params": [{"commitment": "processed"}]
    });

    let response = client.post(rpc_url).json(&request).send().await?;
    let json_value: Value = response.json().await?;

    if let Some(slot) = json_value.get("result").and_then(|v| v.as_u64()) {
        Ok(slot)
    } else if let Some(error) = json_value.get("error") {
        Err(anyhow::anyhow!("getSlot error: {}", error))
    } else {
        Err(anyhow::anyhow!("Failed to get slot"))
    }
}

async fn get_block_time(client: &Client, rpc_url: &str, slot: u64) -> Result<Option<i64>> {
    let request = json!({
        "jsonrpc": "2.0",
        "id": 1,
        "method": "getBlockTime",
        "params": [slot]
    });

    let response = client.post(rpc_url).json(&request).send().await?;
    let json_value: Value = response.json().await?;

    if let Some(error) = json_value.get("error") {
        return Ok(None);
    }

    if let Some(result) = json_value.get("result") {
        if result.is_null() {
            Ok(None)
        } else {
            Ok(result.as_i64())
        }
    } else {
        Ok(None)
    }
}

File: nodejs/latency-calculator.js
#!/usr/bin/env node
import { program } from 'commander';
import fetch from 'node-fetch';
import WebSocket from 'ws';

const METHODS = ['rpc', 'grpc', 'websocket'];

class LatencyMeasurement {
    constructor(slot, blockTime, receivedTime, latencyMs) {
        this.slot = slot;
        this.blockTime = blockTime;
        this.receivedTime = receivedTime;
        this.latencyMs = latencyMs;
    }
}

program
    .name('latency-calculator')
    .description('Calculate average latency for RPC or gRPC over specified number of blocks')
    .requiredOption('--method <method>', `Method to test: ${METHODS.join(', ')}`)
    .requiredOption('--endpoint <url>', 'Endpoint URL')
    .option('--api-key <key>', 'API key (for gRPC)')
    .requiredOption('--blocks <number>', 'Number of blocks to calculate average latency', parseInt)
    .option('--verbose', 'Verbose logging', false);

program.parse();
const args = program.opts();

if (!METHODS.includes(args.method)) {
    console.error(`Error: Method must be one of: ${METHODS.join(', ')}`);
    process.exit(1);
}

console.log('🚀 Latency Calculator');
console.log(`Method: ${args.method}`);
console.log(`Endpoint: ${args.endpoint}`);
console.log(`Target blocks: ${args.blocks}`);
console.log();

async function main() {
    try {
        let measurements;
        
        switch (args.method) {
            case 'rpc':
                measurements = await measureRpcLatency(args);
                break;
            case 'grpc':
                measurements = await measureGrpcLatency(args);
                break;
            case 'websocket':
                measurements = await measureWebsocketLatency(args);
                break;
            default:
                throw new Error(`Unsupported method: ${args.method}`);
        }

        printResults(measurements, args);
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

async function measureRpcLatency(args) {
    const measurements = [];
    let processedBlocks = 0;

    console.log('📡 Starting RPC latency measurement...');
    console.log('Slot       | Block Time    | Received Time | Latency   | Status');
    console.log('-'.repeat(70));

    let lastSlot = await getLatestSlot(args.endpoint);

    while (processedBlocks < args.blocks) {
        try {
            const currentSlot = await getLatestSlot(args.endpoint);
            
            if (currentSlot > lastSlot) {
                const blockTime = await getBlockTime(args.endpoint, currentSlot);
                
                if (blockTime !== null) {
                    const receivedTime = Date.now();
                    const latencyMs = receivedTime - (blockTime * 1000);

                    if (latencyMs > 0 && latencyMs < 10000) {
                        const measurement = new LatencyMeasurement(
                            currentSlot,
                            blockTime,
                            receivedTime,
                            latencyMs
                        );

                        const status = getLatencyStatus(latencyMs);
                        
                        console.log(
                            `${currentSlot.toString().padEnd(10)} | ${blockTime.toString().padEnd(12)} | ${Math.floor(receivedTime / 1000).toString().padEnd(12)} | ${latencyMs.toString().padEnd(9)}ms | ${status}`
                        );

                        measurements.push(measurement);
                        processedBlocks++;

                        if (args.verbose) {
                            console.log(`Progress: ${processedBlocks}/${args.blocks} blocks processed`);
                        }
                    }
                } else if (args.verbose) {
                    console.log(`Block time not available for slot ${currentSlot}`);
                }
                
                lastSlot = currentSlot;
            }
        } catch (error) {
            if (args.verbose) {
                console.error('Error getting latest slot:', error.message);
            }
        }

        await sleep(500);
    }

    return measurements;
}

async function measureGrpcLatency(args) {
    const apiKey = args.apiKey || process.env.HELIUS_API_KEY;
    if (!apiKey) {
        throw new Error('API key required for gRPC method');
    }

    const { subscribe, CommitmentLevel } = await import('helius-laserstream');
    
    const measurements = [];
    let processedBlocks = 0;

    console.log('📡 Starting gRPC latency measurement...');
    console.log('Slot       | Block Time    | Received Time | Latency   | Status');
    console.log('-'.repeat(70));

    const config = {
        apiKey: apiKey,
        endpoint: args.endpoint,
    };

    const request = {
        blocksMeta: {
            all: {}
        },
        commitment: CommitmentLevel.PROCESSED,
        transactions: {},
        accounts: {},
        slots: {},
        transactionsStatus: {},
        blocks: {},
        entry: {},
        accountsDataSlice: [],
    };

    return new Promise(async (resolve, reject) => {
        let completed = false;
        
        const timeoutId = setTimeout(() => {
            if (!completed) {
                completed = true;
                if (processedBlocks === 0) {
                    reject(new Error('No blocks received within timeout period'));
                } else {
                    console.log(`Timeout reached. Collected ${processedBlocks} blocks.`);
                    resolve(measurements);
                }
            }
        }, 60000);

        try {
            await subscribe(
                config,
                request,
                async (data) => {
                    if (completed) return;
                    
                    const receivedTime = Date.now();
                    
                    if (data.blockMeta && processedBlocks < args.blocks) {
                        const slot = data.blockMeta.slot;
                        const blockTime = data.blockMeta.blockTime?.timestamp;
                        
                        if (blockTime) {
                            const latencyMs = receivedTime - (blockTime * 1000);

                            if (latencyMs > 0 && latencyMs < 10000) {
                                const measurement = new LatencyMeasurement(
                                    slot,
                                    blockTime,
                                    receivedTime,
                                    latencyMs
                                );

                                const status = getLatencyStatus(latencyMs);
                                
                                console.log(
                                    `${slot.toString().padEnd(10)} | ${blockTime.toString().padEnd(12)} | ${Math.floor(receivedTime / 1000).toString().padEnd(12)} | ${latencyMs.toString().padEnd(9)}ms | ${status}`
                                );

                                measurements.push(measurement);
                                processedBlocks++;

                                if (args.verbose) {
                                    console.log(`Progress: ${processedBlocks}/${args.blocks} blocks processed`);
                                }

                                if (processedBlocks >= args.blocks && !completed) {
                                    completed = true;
                                    clearTimeout(timeoutId);
                                    resolve(measurements);
                                }
                            }
                        }
                    }
                },
                async (error) => {
                    if (!completed) {
                        completed = true;
                        clearTimeout(timeoutId);
                        reject(error);
                    }
                }
            );
        } catch (error) {
            if (!completed) {
                completed = true;
                clearTimeout(timeoutId);
                reject(error);
            }
        }
    });
}

async function measureWebsocketLatency(args) {
    const measurements = [];
    let processedBlocks = 0;

    console.log('📡 Starting WebSocket latency measurement...');
    console.log('Slot       | Block Time    | Received Time | Latency   | Status');
    console.log('-'.repeat(70));

    let wsUrl = args.endpoint;
    if (wsUrl.startsWith('https://')) {
        wsUrl = wsUrl.replace('https://', 'wss://');
    } else if (wsUrl.startsWith('http://')) {
        wsUrl = wsUrl.replace('http://', 'ws://');
    } else if (!wsUrl.startsWith('ws://') && !wsUrl.startsWith('wss://')) {
        wsUrl = `wss://${wsUrl}`;
    }

    return new Promise((resolve, reject) => {
        const ws = new WebSocket(wsUrl);
        let subscriptionConfirmed = false;
        
        const timeout = setTimeout(() => {
            if (!subscriptionConfirmed) {
                ws.close();
                reject(new Error('WebSocket subscription timeout'));
            } else {
                console.log('No new blocks received in 30 seconds, continuing...');
            }
        }, 30000);

        ws.on('open', () => {
            const subscription = {
                jsonrpc: '2.0',
                id: 1,
                method: 'blockSubscribe',
                params: [
                    'all',
                    {
                        commitment: 'processed',
                        encoding: 'json',
                        transactionDetails: 'none',
                        rewards: false
                    }
                ]
            };
            
            ws.send(JSON.stringify(subscription));
        });

        ws.on('message', (data) => {
            const receivedTime = Date.now();
            
            try {
                const jsonMsg = JSON.parse(data.toString());
                
                if (args.verbose) {
                    console.log('Received WebSocket message:', JSON.stringify(jsonMsg, null, 2));
                }

                if (jsonMsg.params && jsonMsg.params.result) {
                    const result = jsonMsg.params.result;
                    const value = result.value;
                    
                    if (value && value.block) {
                        const slot = value.slot || (value.block.parentSlot ? value.block.parentSlot + 1 : 0);
                        const blockTime = value.block.blockTime;
                        
                        if (blockTime) {
                            const latencyMs = receivedTime - (blockTime * 1000);

                            if (latencyMs > 0 && latencyMs < 10000) {
                                const measurement = new LatencyMeasurement(
                                    slot,
                                    blockTime,
                                    receivedTime,
                                    latencyMs
                                );

                                const status = getLatencyStatus(latencyMs);
                                
                                console.log(
                                    `${slot.toString().padEnd(10)} | ${blockTime.toString().padEnd(12)} | ${Math.floor(receivedTime / 1000).toString().padEnd(12)} | ${latencyMs.toString().padEnd(9)}ms | ${status}`
                                );

                                measurements.push(measurement);
                                processedBlocks++;

                                if (args.verbose) {
                                    console.log(`Progress: ${processedBlocks}/${args.blocks} blocks processed`);
                                }

                                if (processedBlocks >= args.blocks) {
                                    clearTimeout(timeout);
                                    ws.close();
                                    resolve(measurements);
                                }
                            }
                        }
                    }
                } else if (jsonMsg.result !== undefined) {
                    subscriptionConfirmed = true;
                    if (args.verbose) {
                        console.log('WebSocket subscription confirmed');
                    }
                }
            } catch (error) {
                if (args.verbose) {
                    console.error('Error parsing WebSocket message:', error.message);
                }
            }
        });

        ws.on('close', () => {
            clearTimeout(timeout);
            if (processedBlocks < args.blocks) {
                console.log('WebSocket connection closed before collecting all measurements');
            }
            resolve(measurements);
        });

        ws.on('error', (error) => {
            clearTimeout(timeout);
            reject(error);
        });
    });
}

function printResults(measurements, args) {
    if (measurements.length === 0) {
        console.log('❌ No measurements collected');
        return;
    }

    const latencies = measurements.map(m => m.latencyMs);
    const total = latencies.reduce((sum, l) => sum + l, 0);
    const avg = total / latencies.length;
    
    const sortedLatencies = [...latencies].sort((a, b) => a - b);
    
    const min = sortedLatencies[0];
    const max = sortedLatencies[sortedLatencies.length - 1];
    const median = sortedLatencies[Math.floor(sortedLatencies.length / 2)];
    const p95Idx = Math.floor(sortedLatencies.length * 0.95);
    const p95 = sortedLatencies[Math.min(p95Idx, sortedLatencies.length - 1)];
    const p99Idx = Math.floor(sortedLatencies.length * 0.99);
    const p99 = sortedLatencies[Math.min(p99Idx, sortedLatencies.length - 1)];

    const excellent = latencies.filter(l => l < 500).length;
    const good = latencies.filter(l => l >= 500 && l < 1000).length;
    const fair = latencies.filter(l => l >= 1000 && l < 2000).length;
    const slow = latencies.filter(l => l >= 2000).length;

    console.log();
    console.log('📊 Latency Results Summary');
    console.log('='.repeat(50));
    console.log(`Method:             ${args.method}`);
    console.log(`Endpoint:           ${args.endpoint}`);
    console.log(`Blocks processed:   ${measurements.length}`);
    console.log(`Average latency:    ${avg.toFixed(1)}ms`);
    console.log(`Min latency:        ${min}ms`);
    console.log(`Max latency:        ${max}ms`);
    console.log(`Median latency:     ${median}ms`);
    console.log(`95th percentile:    ${p95}ms`);
    console.log(`99th percentile:    ${p99}ms`);
    console.log();

    console.log('⚡ Performance Distribution:');
    console.log(`🟢 Excellent (<500ms):   ${excellent}/${measurements.length} (${((excellent / measurements.length) * 100).toFixed(1)}%)`);
    console.log(`🟡 Good (500-1000ms):    ${good}/${measurements.length} (${((good / measurements.length) * 100).toFixed(1)}%)`);
    console.log(`🟠 Fair (1000-2000ms):   ${fair}/${measurements.length} (${((fair / measurements.length) * 100).toFixed(1)}%)`);
    console.log(`🔴 Slow (>2000ms):       ${slow}/${measurements.length} (${((slow / measurements.length) * 100).toFixed(1)}%)`);
    console.log();

    console.log('🎯 Overall Assessment:');
    if (avg < 500) {
        console.log('✅ EXCELLENT - Very fast latency!');
    } else if (avg < 1000) {
        console.log('🟡 GOOD - Acceptable latency for most use cases');
    } else if (avg < 2000) {
        console.log('🟠 FAIR - Moderate latency, consider optimization');
    } else {
        console.log('🔴 SLOW - High latency, investigate network/provider issues');
    }
}

function getLatencyStatus(latencyMs) {
    if (latencyMs < 500) {
        return '🟢 EXCELLENT';
    } else if (latencyMs < 1000) {
        return '🟡 GOOD';
    } else if (latencyMs < 2000) {
        return '🟠 FAIR';
    } else {
        return '🔴 SLOW';
    }
}

async function getLatestSlot(rpcUrl) {
    const request = {
        jsonrpc: '2.0',
        id: 1,
        method: 'getSlot',
        params: [{ commitment: 'processed' }]
    };

    const response = await fetch(rpcUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(request)
    });

    const jsonValue = await response.json();

    if (jsonValue.result !== undefined) {
        return jsonValue.result;
    } else if (jsonValue.error) {
        throw new Error(`getSlot error: ${JSON.stringify(jsonValue.error)}`);
    } else {
        throw new Error('Failed to get slot');
    }
}

async function getBlockTime(rpcUrl, slot) {
    const request = {
        jsonrpc: '2.0',
        id: 1,
        method: 'getBlockTime',
        params: [slot]
    };

    const response = await fetch(rpcUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(request)
    });

    const jsonValue = await response.json();

    if (jsonValue.error) {
        return null;
    }

    return jsonValue.result;
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

main().catch(console.error);